api-server/api-server-common/src/storage/impls/postgres/queries.rs

```diff
@@ -1096,6 +1119,33 @@ impl<'a, 'b> QueryFromConnection<'a, 'b> {
         let delegation_id = Address::new(chain_config, delegation_id)
             .map_err(|_| ApiServerStorageError::AddressableError)?;
 
+        // copy all delegations for the same pool_id as the new delegation with the new
+        // block_height, and ignore duplicates
+        self.tx.execute(
+            r#"
+            WITH latest_block_height AS (
+                SELECT MAX(block_height) AS max_block_height
+                FROM ml.delegations
+                WHERE pool_id = $2
+            )
+            INSERT INTO ml.delegations (delegation_id, block_height, creation_block_height, pool_id, balance, next_nonce, spend_destination)
+            SELECT delegation_id,
+                   $1,
+                   creation_block_height,
+                   pool_id,
+                   balance,
+                   next_nonce,
+                   spend_destination
+            FROM ml.delegations
+            WHERE pool_id = $2
+              AND block_height = (SELECT max_block_height FROM latest_block_height)
+            ON CONFLICT DO NOTHING;
+            "#,
+            &[&height, &pool_id.as_str()]
+        )
+        .await
+        .map_err(|e| ApiServerStorageError::LowLevelStorageError(e.to_string()))?;
```

Comment:
This looks hacky:

1. Copying all the delegations should probably be done once per each block height and not every time a delegation is modified (which may happen multiple times per block).

2. Copying itself feels hacky. Have you investigated other approaches? E.g. what if we had an additional table "delegation_id -> max_block_height", would it help speed up `get_pool_delegation_shares`?

Reply:
Btw, regarding reorgs - can't we just update the "delegation_id -> max_block_height" table in `del_delegations_above_height`? IMO we should be able to do it in the same sql query that deletes the delegations, so it should be relatively efficient. Or am I missing something?

Reply:
added the cache table